tests/functional/test_fast_deps.py

```diff
@@ -101,3 +103,31 @@ def test_hash_mismatch(script: PipTestEnvironment, tmp_path: pathlib.Path) -> No
         expect_error=True,
     )
     assert "DO NOT MATCH THE HASHES" in result.stderr
+
+
+@mark.network
+def test_hash_mismatch_existing_download(
+    script: PipTestEnvironment, tmp_path: pathlib.Path
+) -> None:
+    reqs = tmp_path / "requirements.txt"
+    reqs.write_text("idna==2.10")
+    dl_dir = tmp_path / "downloads"
+    dl_dir.mkdir()
+    idna_wheel = dl_dir / "idna-2.10-py2.py3-none-any.whl"
+    idna_wheel.write_text("asdf")
+    result = script.pip(
+        "download",
+        "--use-feature=fast-deps",
+        "-r",
+        str(reqs),
+        "-d",
+        str(dl_dir),
+        allow_stderr_warning=True,
+    )
+    assert re.search(
+        r"WARNING: Previously-downloaded file.*has bad hash", result.stderr
+    )
+    assert (
+        hash_file(str(idna_wheel))[0].hexdigest()
+        == "b97d804b1e9b523befed77c48dacec60e6dcb0b5391d57af6a65a312a90648c0"
```

Comment:
If this is the hash of the content "asdf", then (a) let's make that clear, and (b) can we not calculate it, so that we only have to change one place? Something like

```
dummy_content = "asdf"
dummy_content_hash = sha256(dummy_content.encode("utf-8")).hexdigest()
```

Although it appears that magic number is *not* the hash of "asdf"... So I'm confused.

Reply:
This is the hash of the actual wheel after it was downloaded, to replace the incorrect value. I have added a comment to this effect.

Reply:
I'm a little concerned about hard-coding the hash of a wheel we download from PyPI. I guess it should be OK, as wheels on PyPI are immutable, but I feel vaguely uncomfortable about this. It would be better if we could test this the way we normally do using our local test data.

But I may be overthinking this. Feel free to leave this as is if you feel it's OK as it stands.